### ISSUE TITLE
Loopy Execution in arraycontext/arraycontext/fake_nump.py

### ISSUE THREAD

**Issue by mitkotak** [inducer/arraycontext#70]
Would it be helpful to move this [loopy execution](https://github.com/inducer/arraycontext/blob/112ada652122d37586386cd7e61ec3b6097e85f3/arraycontext/fake_numpy.py#L149-L165) from BaseFakeNumpySpace to PyOpenCLArrayContext so that it looks something like                                      [this](https://github.com/inducer/arraycontext/blob/112ada652122d37586386cd7e61ec3b6097e85f3/arraycontext/impl/pytato/fake_numpy.py#L53-L62) and doesn't cause issues with arraycontext's that don't yet support Loopy implementation such as [pycuda arraycontext](https://github.com/inducer/loopy/pull/474) ?

**Comment by kaushikcfd** [comment]
> so that it looks something like this

We could simply leave `BaseFakeNumpyNamespace._getattr__` undefined.

**Comment by alexfikl** [comment]
> > so that it looks something like this
> 
> We could simply leave `BaseFakeNumpyNamespace._getattr__` undefined.

Yeah, agreed. Maybe move the existing scalar ufunc logic into `arraycontext.loopy` so that other array contexts can use it if they want?

**Comment by kaushikcfd** [comment]
>  Maybe move the existing scalar ufunc logic into arraycontext.loopy so that other array contexts can use it if they want?

Yep, agreed.

> Or is that very PyOpenCL specific?

I feel like it is. At the very least, forcing the [LoopKernel.target](https://documen.tician.de/loopy/ref_kernel.html?highlight=loopkernel#loopy.LoopKernel.target) to be PyOpenCLTarget seems like a bug.

**Comment by inducer** [comment]
I'm not fundamentally opposed to this. However I'm still trying to understand the motivation. `call_loopy` is part of the documented interface of an array context, so we should be able to rely on it. Did this come from trying to wrap JAX into an array context?

**Comment by kaushikcfd** [comment]
> Did this come from trying to wrap JAX into an array context?

This popped up while @mitkotak was working on a PyCUDAArrayContext and which currently doesn't have `call_loopy` support.

> `call_loopy` is part of the documented interface of an array context, so we should be able to rely on 

Yep agreed. Because of this we weren't completely sure about this too and hence wanted to run this by and see if the `BaseFakeNumpyNamespace.__getattr__` is still a good fallback.

**Comment by inducer** [comment]
Thanks for explaining. I'd be OK with doing what @alexfikl proposes, which is pushing this off into a `loopy` mix-in-ish class that array contexts can (but don't have to) use.
